plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
+    | Emit t (Contract t i a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Done -> Done
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> case ca of
+            Done       -> Done
+            Emit t c   -> Emit t (Pure f <*> c)
+            Pure a     -> Pure (f a)
+            Waiting f' -> Waiting $ \i' -> fmap f (f' i')
+        Waiting f ->
+            case ca of
+                Done -> Done
+                Emit t c -> Emit t (Waiting f <*> c)
+                Pure a' -> Waiting $ \i' -> fmap (\f' -> f' a') (f i')
+                Waiting f' ->
+                    Waiting $ \i' -> f i' <*> f' i'
+
+-- The monad instance sequentialises the 'Waiting' operations
+instance Monad (Contract t i) where
+    c >>= f = case c of
+        Waiting f' -> Waiting (f' >=> f)
+        Done       -> Done
+        Pure a     -> f a
+        Emit t c'  -> Emit t (c' >>= f)
+
+instance Semigroup a => Semigroup (Contract t i a) where
+    (<>) = liftA2 (<>)
```

Comment:
This waits for both sides to finish. Another option would be to select the contrat that finishes first (ie. match on the arguments and return the first `Pure` we see). This might actually be nicer - it has the benefit of not needing a `Semigroup a`, we could have `Done` as a unit and it would also work better with recursion (for contracts that have some internal state)

Reply:
So as I understand it, this waits for both to finish and then combines them? That seems reasonable. We could provide the "racing" one as the `Alternative` instance, perhaps?